File: my-folder/1371-minimum-remove-to-make-valid-parentheses/solution.java
class Solution {
  public String minRemoveToMakeValid(String s) {
    char[] sArr = s.toCharArray();
    
    int openBracketsCount = 0;
    
    for (int i = 0; i < sArr.length; i++) {
      if (sArr[i] == '(') {
        openBracketsCount++;
      } else if (sArr[i] == ')') {
        if (openBracketsCount > 0) {
          openBracketsCount--;
        } else {
          sArr[i] = '*';
        }
      }
    }
    
    for (int i = sArr.length - 1; i >= 0; i--) {
      if (openBracketsCount == 0) {
        break;
      }
      
      if (sArr[i] == '(') {
        sArr[i] = '*';
        openBracketsCount--;
      }
    }
    
    StringBuilder strBld = new StringBuilder();
    for (int i = 0; i < sArr.length; i++) {
      if (sArr[i] != '*') {
        strBld.append(sArr[i]);
      }
    }
    
    return strBld.toString();
  }
}

File: my-folder/1371-minimum-remove-to-make-valid-parentheses/solution.kt
class Solution {
    fun minRemoveToMakeValid(s: String): String {
      val set = HashSet<Int>()
      var counter = 0
      
      for (i in 0 until s.length) {
        if (s[i] == '(') ++counter
        if (s[i] == ')') {
          if (counter > 0) --counter 
          else set.add(i)
        }
      }
      
      counter = 0
      for (i in s.length-1 downTo 0) {
        if (s[i] == ')') ++counter
        if (s[i] == '(') {
          if (counter > 0) --counter 
          else set.add(i)
        }
      }
      
      val sb = StringBuilder()
      for (i in 0 until s.length) {
        if (!set.contains(i)) sb.append(s[i])
      }
      
      return sb.toString()
    }
}
